piecedirectory/doctor.go

```diff
@@ -98,12 +100,77 @@ func (d *Doctor) Run(ctx context.Context) {
 }
 
 func (d *Doctor) checkPiece(ctx context.Context, pieceCid cid.Cid, lu *sectorstatemgr.SectorStateUpdates) error {
+	defer func(start time.Time) { log.Debugw("checkPiece processing", "took", time.Since(start)) }(time.Now())
+
+	////////////////////////////////////////////////
+	// Check if piece belongs to an active sector
+	////////////////////////////////////////////////
 	md, err := d.store.GetPieceMetadata(ctx, pieceCid)
 	if err != nil {
 		return fmt.Errorf("failed to get piece %s from local index directory: %w", pieceCid, err)
 	}
 
+	lacksActiveSector := true // check whether the piece is present in active sector
+
+	var chainDealIds []abi.DealID
+	for _, dl := range md.Deals {
+		mid, err := address.IDFromAddress(dl.MinerAddr)
+		if err != nil {
+			return err
+		}
+
+		sectorID := abi.SectorID{
+			Miner:  abi.ActorID(mid),
+			Number: dl.SectorID,
+		}
+
+		// check if we have an active sector
+		if _, ok := lu.ActiveSectors[sectorID]; ok {
+			lacksActiveSector = false
+			chainDealIds = append(chainDealIds, dl.ChainDealID)
+		}
+	}
+
+	if lacksActiveSector {
+		doclog.Debugw("ignoring and unflagging piece as it is not present in an active sector", "piece", pieceCid)
+
+		err = d.store.UnflagPiece(ctx, pieceCid)
+		if err != nil {
+			return fmt.Errorf("failed to unflag piece %s: %w", pieceCid, err)
+		}
+		return nil
+	}
+
+	/////////////////////////////////////////////////////////////////
+	// Check that Deal is actually on-chain for the active sectors
+	/////////////////////////////////////////////////////////////////
+	head, err := d.fullnodeApi.ChainHead(ctx)
+	if err != nil {
+		return err
+	}
+
+	found := false
+	for _, dealId := range chainDealIds {
+		_, err := d.fullnodeApi.StateMarketStorageDeal(ctx, dealId, head.Key())
```

Comment:
We might also want to consider passing the StateMarketStorageDeal function as a parameter to checkPiece, so that it's easier to test (without having to use mocks) - take a look at testCheckPieces in doctor_test.go